tss-esapi/src/structures/attest.rs

```diff
@@ -0,0 +1,173 @@
+
+    fn try_from(tpms_attest: TPMS_ATTEST) -> Result<Self> {
+        if tpms_attest.magic != TPM2_GENERATED_VALUE {
+            return Err(Error::local_error(WrapperErrorKind::InvalidParam));
+        }
+
+        let attestation_type = AttestationType::try_from(tpms_attest.type_)?;
+        Ok(Attest {
+            attestation_type,
+            qualified_signer: Name::try_from(tpms_attest.qualifiedSigner)?,
+            extra_data: Data::try_from(tpms_attest.extraData)?,
+            clock_info: ClockInfo::try_from(tpms_attest.clockInfo)?,
+            firmware_version: tpms_attest.firmwareVersion,
+            attested: match attestation_type {
+                AttestationType::Certify => AttestInfo::Certify {
+                    info: unsafe { tpms_attest.attested.certify }.try_into()?,
+                },
+                AttestationType::Quote => AttestInfo::Quote {
+                    info: unsafe { tpms_attest.attested.quote }.try_into()?,
+                },
+                AttestationType::SessionAudit => AttestInfo::SessionAudit {
+                    info: unsafe { tpms_attest.attested.sessionAudit }.try_into()?,
+                },
+                AttestationType::CommandAudit => AttestInfo::CommandAudit {
+                    info: unsafe { tpms_attest.attested.commandAudit }.try_into()?,
+                },
+                AttestationType::Time => AttestInfo::Time {
+                    info: unsafe { tpms_attest.attested.time }.try_into()?,
+                },
+                AttestationType::Creation => AttestInfo::Creation {
+                    info: unsafe { tpms_attest.attested.creation }.try_into()?,
+                },
+                AttestationType::Nv => AttestInfo::Nv {
+                    info: unsafe { tpms_attest.attested.nv }.try_into()?,
+                },
+                AttestationType::NvDigest => {
+                    error!("NvDigest attestation type is currently not supported");
+                    return Err(Error::local_error(WrapperErrorKind::UnsupportedParam));
+                }
+            },
+        })
+    }
+}
+
+impl Marshall for Attest {
+    const BUFFER_SIZE: usize = std::mem::size_of::<TPMS_ATTEST>();
+
+    fn marshall(&self) -> Result<Vec<u8>> {
+        let mut buffer = Vec::with_capacity(Self::BUFFER_SIZE);
+        let mut offset: crate::tss2_esys::size_t = 0;
+
+        let ret = Error::from_tss_rc(unsafe {
+            crate::tss2_esys::Tss2_MU_TPMS_ATTEST_Marshal(
+                &self.clone().into(),
+                buffer.as_mut_ptr(),
+                Self::BUFFER_SIZE.try_into().map_err(|e| {
+                    error!("Failed to convert size of buffer to TSS size_t type: {}", e);
+                    Error::local_error(WrapperErrorKind::InvalidParam)
+                })?,
+                &mut offset,
+            )
+        });
+
+        if !ret.is_success() {
+            return Err(ret);
+        }
+
+        unsafe { buffer.set_len(Self::BUFFER_SIZE) };
```

Comment:
I don't think this is correct - the number of marshalled bytes might be smaller than this and the exact size is given in `offset`. See, for example, the implementation [here](https://github.com/parallaxsecond/rust-tss-esapi/blob/main/tss-esapi/src/abstraction/transient/key_attestation.rs#L91). That `BUFFER_SIZE` is just the max possible value. The [spec](https://trustedcomputinggroup.org/wp-content/uploads/TCG_TSS_Marshaling_Unmarshaling_API_v1p0_r07_pub.pdf) says:

> The `offset` parameter will be updated by the function to the index of the next byte after the marshaled data 
in the `buffer` byte array.

Pointing this out because in the case I link to above I need the buffer to be truncated correctly, otherwise hashing it doesn't result in the correct object name. Other structures might have similar requirements.

Reply:
Yes, you are right I was going to use it like that but for some reason forgot to use the offset parameter.